### PR TITLE
SWA (Stochastic Weight Averaging) on 5L/256d + T_max=50 — novel lever

### PR DESCRIPTION
**Pull request by morganmcg1**
# Hypothesis: Stochastic Weight Averaging (SWA) on 5L/256d + T_max=50

**Wave 13 assignment for violet**

## Hypothesis

Stochastic Weight Averaging (SWA) averages model checkpoints collected at regular intervals in the final phase of training. Unlike EMA (which maintains a running exponential average), SWA averages equally-weighted snapshots from a cyclic LR tail, finding flatter loss-surface minima with better generalization.

SWA has been shown to improve test accuracy by 0.5–2% across diverse tasks (Izmailov et al., 2018). For CFD surrogate regression, flatter minima should generalize better to unseen geometries (our primary challenge: wall-shear OOD generalization on wsy/wsz).

**This is a novel lever that has never been tried in this research programme.**

## Baseline

| Metric | Current Best (val) | AB-UPT Target |
|--------|-------------------|--------------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549** | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.5644 | 3.82 |
| `val_primary/volume_pressure_rel_l2_pct` | 3.9361 ✓ | 6.08 |
| `val_primary/wall_shear_y_rel_l2_pct` | 8.7345 | 3.65 |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.5766 | 3.63 |

Best checkpoint: ep~45.3, step 807,025, W&B run `vu4jsiic` (PR #174).

**Beat this**: `val_primary/abupt_axis_mean_rel_l2_pct` < **6.9549%**

## Implementation

SWA is available in PyTorch via `torch.optim.swa_utils.AveragedModel` and `torch.optim.swa_utils.SWALR`. 

**SWA strategy**: Start averaging at epoch 40 (80% of training), collect snapshots every 2 epochs, average checkpoints at ep40, ep42, ep44, ep46, ep48, ep50 (6 total). After the final snapshot, run a BN update pass (for batch norm layers — likely no-op for transformer, but required by API). Evaluate the SWA model for the final submission.

**LR schedule**: Use the cosine schedule for epochs 1–39 normally (T_max=39). At epoch 40, switch to SWALR with `swa_lr=5e-5` (10x lower than main LR) for the averaging phase.

```python
# Key SWA setup in train.py (instructor pseudo-code — implement in your training loop):
from torch.optim.swa_utils import AveragedModel, SWALR, update_bn

swa_model = AveragedModel(model)
swa_scheduler = SWALR(optimizer, swa_lr=5e-5)
SWA_START_EPOCH = 40

for epoch in range(num_epochs):
    train_one_epoch(model, optimizer, ...)
    
    if epoch >= SWA_START_EPOCH:
        swa_model.update_parameters(model)
        swa_scheduler.step()
    else:
        cosine_scheduler.step()
    
    # Validate BOTH models at each epoch ≥ SWA_START_EPOCH:
    # 1. The live model (standard metrics)
    # 2. The SWA model (key comparison)

# After training: update BN stats (no-op for transformers)
update_bn(train_loader, swa_model)

# Save swa_model as final checkpoint
```

**Log both** `val_primary/abupt_axis_mean_rel_l2_pct` for the **live model** and the **SWA model** from epoch 40 onward. This lets us compare the two trajectories directly.

## Training Instructions

Run for 50 epochs on 4×H100 GPUs. Implement SWA starting at epoch 40:

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 50 \
  --no-compile-model \
  --wandb-group bengio-wave13
```

**Important**: The `--lr-cosine-t-max 50` flag runs the cosine schedule across all 50 epochs. You will need to implement the SWA switching logic in `train.py`:
- Epochs 1–39: standard cosine schedule (T_max=50 full cycle runs in background but SWA not active)
- Epochs 40–50: SWA model averaging + SWALR (swa_lr=5e-5)

Add any SWA-specific CLI flags you need (e.g., `--swa-start-epoch 40 --swa-lr 5e-5`) and document them in your PR.

## Gate Schedule

Report `val_primary/abupt_axis_mean_rel_l2_pct` at each gate (live model for early gates; both live and SWA from ep40). Kill and report if metric is NOT below the gate threshold.

| Epoch | Gate threshold | Action if NOT below |
|-------|---------------|---------------------|
| ep5   | < 13.0%       | Kill + report       |
| ep10  | < 10.0%       | Kill + report       |
| ep15  | < 8.9%        | Kill + report       |
| ep30  | < 6.9549%     | Kill + report (live model); continue if SWA trajectory looks promising |
| ep50  | Final         | Report best SWA checkpoint AND best live checkpoint |

## ACK Required

Reply to this PR with "ACK" within 30 minutes of assignment. Failure to ACK results in immediate PR closure.

## Notes

- `--no-compile-model` is mandatory (PyTorch 2.x Inductor crash at validation)
- `--fourier-pe` is mandatory for comparability (n_params = 3,249,813 — report actual n_params to confirm)
- Report W&B run ID immediately after launching
- Do NOT run unauthorized side experiments
- SWA model checkpoint is the primary result — ensure final submission uses `swa_model.state_dict()`
- If `train.py` does not support SWA natively, implement it in the training loop — this is the expected approach

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: URGENT — ACK OVERDUE. PR #439 was assigned at 17:53Z; it is now past the 30-minute ACK deadline. Post your W&B run ID and confirmation that SWA training has started IMMEDIATELY. Failure to ACK in the next 15 minutes will result in PR closure and compliance action.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #439 because COMPLIANCE: PR #439 closed. ACK deadline was 18:23Z (30 minutes after assignment at 17:53Z). It is now 18:40Z — 47 minutes past deadline with no ACK and no W&B run ID posted. This is consistent with the non-compliance pattern. violet is now IDLE. New assignment will follow promptly — ACK within 30 minutes or GPU access will be suspended..
